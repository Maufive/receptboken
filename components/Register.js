import React, { Component } from "react";
import axios from "axios";
import Router from "next/router";
import { LoginForm } from "./styles/ModalStyles";
import { Button, DisabledButton } from "./styles/Button";
import MailIcon from "../svg/mail.svg";
import LockIcon from "../svg/lock.svg";
import ProfileIcon from "../svg/profile.svg";

class RegisterComponent extends Component {
	constructor(props) {
		super();
		this.state = {
			fname: "",
			lname: "",
			email: "",
			password: "",
			confirmPassword: "",
			passwordMatch: false
		};
	}

	login = (email, password) => {
		axios
			.post(`https://receptboken.herokuapp.com/auth/login`, { email, password })
			.then(result => {
				localStorage.setItem("jwtToken", result.data.token);
				console.log("JWT Token Set");
				console.log(result.data);
				console.log(this.props.setUser);
				this.props.setUser(result.data.user);
				this.props.setMessage(
					"success",
					`üëã V√§lkommen ${result.data.user.fname}!`
				);
				Router.push("/");
			})
			.catch(error => {
				console.log(error);
				console.log(error.response.data.message);
				this.setState({ error: error.response.data.message });
			});
	};

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	password = async e => {
		await this.setState({ password: e.target.value });
		this.checkPassword();
	};

	confirmPassword = async e => {
		await this.setState({ confirmPassword: e.target.value });
		this.checkPassword();
	};

	checkPassword = () => {
		const { password, confirmPassword } = this.state;
		if (password === confirmPassword && password.length) {
			this.setState({ passwordMatch: true });
		} else {
			this.setState({ passwordMatch: false });
		}
	};

	onSubmit = e => {
		e.preventDefault();

		const { email, password, passwordMatch, fname, lname } = this.state;
		if (passwordMatch) {
			axios
				.post("https://receptboken.herokuapp.com/auth/register", {
					email,
					password,
					fname,
					lname
				})
				.then(result => {
					this.login(email, password);
					this.setState({
						email: "",
						password: "",
						fname: "",
						lname: ""
					});
				})
				.catch(error => {
					console.log(error.response);
					this.props.setMessage("danger", error.response.data);
				});
		}

		if (!passwordMatch) {
			this.props.setMessage("danger", "L√∂senorden matchar inte!");
			throw new Error("urk p√• burk");
		}
	};

	render() {
		return (
			<div style={{ width: "0 auto" }}>
				<LoginForm onSubmit={this.onSubmit}>
					<h2>Registrera dig h√§r:</h2>
					<div>
						<input
							type="email"
							name="email"
							id="email"
							value={this.state.email}
							onChange={this.saveToState}
							required
						/>
						<label htmlFor="email" for="email">
							<MailIcon /> Email:
						</label>
					</div>
					<div>
						<input
							type="text"
							name="fname"
							value={this.state.fname}
							onChange={this.saveToState}
							id="fname"
							required
						/>
						<label htmlFor="name">
							<ProfileIcon /> F√∂rnamn:
						</label>
					</div>
					<div>
						<input
							type="text"
							name="lname"
							id="lname"
							value={this.state.lname}
							onChange={this.saveToState}
							required
						/>
						<label htmlFor="lname">
							<ProfileIcon /> Efternamn:
						</label>
					</div>
					<div>
						<input
							type="password"
							name="password"
							id="password"
							value={this.state.password}
							onChange={this.password}
							required
						/>
						<label htmlFor="password">
							{this.state.passwordMatch ? (
								<LockIcon style={{ fill: "#15BD76" }} />
							) : (
								<LockIcon />
							)}{" "}
							L√∂senord:
						</label>
					</div>
					<div>
						<input
							type="password"
							name="password"
							id="confirmPassword"
							value={this.state.confirmPassword}
							onChange={this.confirmPassword}
							required
						/>
						<label htmlFor="password" for="confirmPassword">
							{this.state.passwordMatch ? (
								<LockIcon style={{ fill: "#15BD76" }} />
							) : (
								<LockIcon />
							)}{" "}
							Bekr√§fta L√∂senord:
						</label>
					</div>
					{this.state.passwordMatch ? (
						<Button fullWidth primary type="submit" id="submit-button">
							Registrera ‚Üí
						</Button>
					) : (
						<DisabledButton disabled fullWidth>
							Registrera ‚Üí
						</DisabledButton>
					)}
				</LoginForm>
			</div>
		);
	}
}

export default RegisterComponent;
