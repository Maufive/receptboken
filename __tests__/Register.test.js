import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import Register from "../components/Register";

const userInput = {
	email: "test@gmail.com",
	fname: "Niklas",
	lname: "Albinsson",
	password: "password123",
	confirmPassword: "password123",
	passwordMatch: true
};

describe("<Register />", () => {
	let wrapper;
	beforeEach(() => (wrapper = shallow(<Register />)));
	// renders to the page
	it("renders to the page and matches snapshot", () => {
		expect(wrapper.find("ModalStyles__LoginForm").exists()).toBe(true);
		expect(toJSON(wrapper)).toMatchSnapshot();
	});

	// renders all the inputs and handles the change-event
	describe("Email input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#email").length).toEqual(1);
		});

		it("handles the onChange-event on the input", () => {
			wrapper.find("#email").simulate("change", {
				target: { name: "email", value: "test@gmail.com" }
			});
		});
	});

	describe("Firstname input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#fname").length).toEqual(1);
		});

		it("handles the onChange-event on the input", () => {
			wrapper.find("#fname").simulate("change", {
				target: { name: "fname", value: "Niklas" }
			});
		});
	});

	describe("Lastname input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#lname").length).toEqual(1);
		});

		it("handles the onChange-event on the input", () => {
			wrapper.find("#lname").simulate("change", {
				target: { name: "lname", value: "Albinsson" }
			});
		});
	});

	describe("Password input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#password").length).toEqual(1);
		});

		it("handles the onChange-event on the input", () => {
			wrapper.find("#password").simulate("change", {
				target: { name: "password", value: "password123" }
			});
		});
	});

	describe("ConfirmPassword input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#confirmPassword").length).toEqual(1);
		});

		it("handles the onChange-event on the input", () => {
			wrapper.find("#confirmPassword").simulate("change", {
				target: { name: "confirmPassword", value: "password123" }
			});
		});
	});

	// handles the submitevent
	describe("Form submitting", () => {
		it("throws an error without input", () => {
			const mockFn = {
				preventDefault: jest.fn()
			};

			expect(() =>
				wrapper.find("ModalStyles__LoginForm").simulate("submit", mockFn)
			).toThrow();

			expect(mockFn.preventDefault).toHaveBeenCalledTimes(1);
		});

		it("submits the form if the input is correct", () => {
			const mockFn = {
				preventDefault: jest.fn()
			};

			wrapper.setState({ ...userInput });

			wrapper.update();
			expect(
				wrapper.find("ModalStyles__LoginForm").simulate("submit", mockFn)
			).toBeTruthy();
			expect(mockFn.preventDefault).toHaveBeenCalledTimes(1);
		});
	});
});
