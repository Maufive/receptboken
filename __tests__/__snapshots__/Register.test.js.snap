// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> renders to the page and matches snapshot 1`] = `
<div
  style={
    Object {
      "width": "0 auto",
    }
  }
>
  <ModalStyles__LoginForm
    onSubmit={[Function]}
  >
    <h2>
      Registrera dig här:
    </h2>
    <div>
      <input
        id="email"
        name="email"
        onChange={[Function]}
        required={true}
        type="email"
        value=""
      />
      <label
        for="email"
        htmlFor="email"
      >
        <MailIcon
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        />
         Email:
      </label>
    </div>
    <div>
      <input
        id="fname"
        name="fname"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <label
        htmlFor="name"
      >
        <ProfileIcon
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        />
         Förnamn:
      </label>
    </div>
    <div>
      <input
        id="lname"
        name="lname"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <label
        htmlFor="lname"
      >
        <ProfileIcon
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        />
         Efternamn:
      </label>
    </div>
    <div>
      <input
        id="password"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
      <label
        htmlFor="password"
      >
        <LockIcon
          viewBox="0 0 58 58"
          xmlns="http://www.w3.org/2000/svg"
        />
         
        Lösenord:
      </label>
    </div>
    <div>
      <input
        id="confirmPassword"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
      <label
        for="confirmPassword"
        htmlFor="password"
      >
        <LockIcon
          viewBox="0 0 58 58"
          xmlns="http://www.w3.org/2000/svg"
        />
         
        Bekräfta Lösenord:
      </label>
    </div>
    <Button__DisabledButton
      disabled={true}
      fullWidth={true}
    >
      Registrera →
    </Button__DisabledButton>
  </ModalStyles__LoginForm>
</div>
`;
